pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -192,6 +196,63 @@ type BMCDetails struct {
 	DisableCertificateVerification bool `json:"disableCertificateVerification,omitempty"`
 }
 
+// HRAIDVolume defines the desired configuration of volume in hardware RAID
+type HRAIDVolume struct {
+	// Size (Integer) of the logical disk to be created in GiB.
+	// If unspecified or set be 0, the maximum capacity of disk will be used for logical disk.
+	// +kubebuilder:validation:Minimum=0
+	SizeGibibytes *int `json:"sizeGibibytes,omitempty"`
+
+	// RAID level for the logical disk. The following levels are supported: 0;1;1+0.
+	// +kubebuilder:validation:Enum="0";"1";"1+0"
+	Level nodes.RAIDLevel `json:"level" required:"true"`
+
+	// Name of the volume. Should be unique within the Node. If not specified, volume name will be auto-generated.
+	// +kubebuilder:validation:MaxLength=64
+	Name string `json:"name,omitempty"`
+
+	// Whether this disk represents rotational storage
+	Rotational *bool `json:"rotational,omitempty"`
+
+	// Integer, number of disks to use for the logical disk. Defaults to minimum number of disks required
+	// for the particular RAID level.
+	// +kubebuilder:validation:Minimum=1
+	NumberOfPhysicalDisks *int `json:"numberOfPhysicalDisks,omitempty"`
+}
+
+// SRAIDVolume defines the desired configuration of volume in software RAID
+type SRAIDVolume struct {
+	// Size (Integer) of the logical disk to be created in GiB.
+	// If unspecified or set be 0, the maximum capacity of disk will be used for logical disk.
+	// +kubebuilder:validation:Minimum=0
+	SizeGibibytes *int `json:"sizeGibibytes,omitempty"`
+
+	// RAID level for the logical disk. The following levels are supported: 0;1;1+0.
+	// +kubebuilder:validation:Enum="0";"1";"1+0"
+	Level nodes.RAIDLevel `json:"level" required:"true"`
+
+	// Name of the volume. Should be unique within the Node. If not specified, volume name will be auto-generated.
+	// +kubebuilder:validation:MaxLength=64
+	Name string `json:"name,omitempty"`
+}
```

Comment:
We need some way to select the physical disks in here, don't we?

If gophercloud only supports the name then I guess we need to decide whether that's sufficient or if we want to improve gophercloud to support the full device hints.
A middle ground could be to create a PhysicalDeviceHints struct that contains only a name; then we could add additional hint types to it later without any backward-incompatible changes to the API.

Reply:
> physical_disks - A list of physical disks to use as read by the RAID interface.
> For software RAID physical_disks is a list of device hints in the same format as used for Specifying the disk for deployment (root device hints). The number of provided hints must match the expected number of backing devices (repeat the same hint if necessary).

I think this means that `physical_disks` is different for `hardware RAID` and `software RAID`, so i think only `hardware RAID` supports `PhysicalDisks`.

I think this is wrong:
```
{
  "logical_disks": [
    {
      "size_gb": 100,
      "raid_level": "5",
      "controller": "software",
      "physical_disks": ["/dev/sda", "/dev/sdb"],
      "is_root_volume": true
    }
  ]
}

```
If we want to add `PhysicalDeviceHints` for `software RAID`, we need to modify `gophercloud`.

Reply:
@zaneb is this resolved?